### PR TITLE
Feature/229-notifications-design

### PR DESCRIPTION
**Pull request by AyTee001**
https://github.com/BinaryStudioAcademy/bsa-2023-leetwars/assets/108694717/6b44e9e4-8a6d-4c46-a504-8c0ad38f4f61

### CODE REVIEW

**Comment by SashaBondarchuk** [comment]
[[BUG#101] [Edit profile page] The edit profile page doesn't match the design in screen resolution 360*640](https://trello.com/c/EAT9WlRR/237-bug101-edit-profile-page-the-edit-profile-page-doesnt-match-the-design-in-screen-resolution-360640)

File: backend/LeetWars.Core/LeetWars.Core.BLL/Services/NotificationService.cs [hunk 1]
```diff
@@ -0,0 +1,115 @@
...
+        {
+            _notificationSenderService = notificationSenderService;
+            _userGetter = userGetter;
+        }
+
+        public async Task CreateNotification(NotificationDto newNotification)
+        {
+            var notification = _mapper.Map<Notification>(newNotification);
+
+            _context.Notifications.Add(notification);
```
Comment (pospyra) [inline backend/LeetWars.Core/LeetWars.Core.BLL/Services/NotificationService.cs]: maybe AddAsync?

File: frontend/src/app/shared/components/header/header.component.ts [hunk 2]
```diff
@@ -40,26 +45,76 @@ export class HeaderComponent extends BaseComponent implements OnInit, OnDestroy
...
+            this.notificationStorage.updateStatusToRead(this.newNotificationsCollection.map(e => e.id))
+                .pipe(takeUntil(this.unsubscribe$))
+                .subscribe({
+                    error: () => {
+                        this.toastrService.showError('Server connection error');
+                        this.router.navigate(['/']);
+                    },
+                });
+        }
+        this.isNotificationsDropdownDisplayed = !this.isNotificationsDropdownDisplayed;
```
Comment (pospyra) [inline frontend/src/app/shared/components/header/header.component.ts]: maybe use ternary operator?

```
public showNotifications() {
    this.isNotificationsDropdownDisplayed
        ? this.readNotifications()
        : this.notificationStorage.updateStatusToRead(this.newNotificationsCollection.map(e => e.id))
            .pipe(takeUntil(this.unsubscribe$))
            .subscribe({
                error: () => {
                    this.toastrService.showError('Server connection error');
                    this.router.navigate(['/']);
                },
            });

    this.isNotificationsDropdownDisplayed = !this.isNotificationsDropdownDisplayed;
}
```

File: frontend/src/app/shared/components/notifications/notifications.component.ts [hunk 3]
```diff
@@ -30,18 +32,24 @@ export class NotificationsComponent extends BaseComponent implements OnInit {
...
     ngOnInit(): void {
         this.getCurrentUser();
-
-        this.notificationService.currentNotifications.subscribe((notifications: INotificationModel[]) => {
-            this.notifications = notifications;
-        });
+        console.log(this.notification);
+        // this.notificationService.currentNotifications.subscribe((notifications: INotificationModel[]) => {
+        //     this.notifications = notifications;
+        // });
```
Comment (pospyra) [inline frontend/src/app/shared/components/notifications/notifications.component.ts]: please remove commented code

**Review by pospyra — CHANGES_REQUESTED** [review]

File: frontend/src/app/shared/components/header/header.component.ts (hunk 2 above)
Comment (AyTee001) [reply frontend/src/app/shared/components/header/header.component.ts]: I guess in this case if-else is better

**Review by pospyra — APPROVED** [review]

File: backend/LeetWars.Core/LeetWars.Core.BLL/Services/NotificationService.cs [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+                var allIds = _context.Users.Select(e => e.Id);
+                foreach(var id in allIds)
+                {
+                    _context.UserNotifications.Add(new UserNotification
+                    {
+                        IsRead = false,
+                        ReceiverId = id,
+                        NotificationId = notification.Id
+                    });
+                }
```
Comment (tatianahutii) [inline backend/LeetWars.Core/LeetWars.Core.BLL/Services/NotificationService.cs]: you can use select and AddRange

File: backend/LeetWars.Core/LeetWars.Core.BLL/Services/NotificationService.cs [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+
+        public async Task CreateNotification(NotificationDto newNotification)
+        {
+            var notification = _mapper.Map<Notification>(newNotification);
+
+            await _context.Notifications.AddAsync(notification);
+
+            await _context.SaveChangesAsync();
+
+            if(newNotification.ReceiverId is null)
```
Comment (tatianahutii) [inline backend/LeetWars.Core/LeetWars.Core.BLL/Services/NotificationService.cs]:
```suggestion
            if (newNotification.ReceiverId is null)
```

File: backend/LeetWars.Core/LeetWars.Core.BLL/Services/NotificationService.cs [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+            {
+                var allIds = _context.Users.Select(e => e.Id);
+                foreach(var id in allIds)
```
Comment (tatianahutii) [inline backend/LeetWars.Core/LeetWars.Core.BLL/Services/NotificationService.cs]:
```suggestion
                foreach (var id in allIds)
```

File: backend/LeetWars.Core/LeetWars.Core.BLL/Services/NotificationService.cs [hunk 7]
```diff
@@ -0,0 +1,108 @@
...
+                }
+            }
+            else
+            {
+                _context.UserNotifications.Add(new UserNotification
+                {
+                    IsRead = false,
+                    ReceiverId = int.Parse(newNotification.ReceiverId),
+                    NotificationId = notification.Id
+                });
```
Comment (tatianahutii) [inline backend/LeetWars.Core/LeetWars.Core.BLL/Services/NotificationService.cs]: and all this if/else can be simplify some thing like
var receiversIds = newNotification.ReceiverId is null
    ? await _context.Users.Select(e => e.Id).ToListAsync()
    : new List<long> { newNotification.ReceiverId };
    
   use Select for creating add notifications and save them all at ones

File: backend/LeetWars.Core/LeetWars.Core.BLL/Services/NotificationService.cs [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        }
+
+        public async Task<ICollection<NotificationDto>> GetNotificationsOfCurrentUserAsync()
+        {
+            var currUserId = _userGetter.CurrentUser?.Id;
+
+            if (currUserId is null)
+            {
+                throw new NotFoundException("User not found");
+            }
```
Comment (tatianahutii) [inline backend/LeetWars.Core/LeetWars.Core.BLL/Services/NotificationService.cs]:
```suggestion
            var currUserId = _userGetter.CurrentUser?.Id ?? throw new NotFoundException("User not found");
```

File: backend/LeetWars.Core/LeetWars.Core.Common/DTO/Friendship/UpdateFriendshipDto.cs [hunk 9]
```diff
@@ -1,11 +1,13 @@
 ﻿using LeetWars.Core.DAL.Enums;
+using System.Text.Json.Serialization;
 
 namespace LeetWars.Core.Common.DTO.Friendship
 {
     public class UpdateFriendshipDto
     {
         public long UserId { get; set; }
         public long FriendshipId { get; set; }
+        [JsonConverter(typeof(JsonStringEnumConverter))]
```
Comment (tatianahutii) [inline backend/LeetWars.Core/LeetWars.Core.Common/DTO/Friendship/UpdateFriendshipDto.cs]: why do we need this?

File: backend/LeetWars.Core/LeetWars.Core.DAL/Migrations/20230928091527_CreateNotifications.cs
Comment (tatianahutii) [inline backend/LeetWars.Core/LeetWars.Core.DAL/Migrations/20230928091527_CreateNotifications.cs:1]: why so big migration?

File: backend/LeetWars.Core/LeetWars.Core.WebAPI/Controllers/NotificationsController.cs [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+            return Ok(notifications);
+        }
+
+        /// <summary>
+        /// Updates status of notifications to read
+        /// </summary>
+        /// <param name="ids">Array of ids of users that have read their messages</param>
+        /// <returns>Status of the operation</returns>
+        [HttpPut]
+        public async Task<ActionResult> UpdateStatusToRead([FromBody] long[] ids)
```
Comment (tatianahutii) [inline backend/LeetWars.Core/LeetWars.Core.WebAPI/Controllers/NotificationsController.cs]:
```suggestion
        public async Task<ActionResult> UpdateStatusToReadAsync([FromBody] long[] ids)
```

File: backend/LeetWars.Core/LeetWars.Core.WebAPI/Controllers/NotificationsController.cs [hunk 11]
```diff
@@ -0,0 +1,45 @@
...
+        {
+            await _notificationService.UpdateStatusToReadByUserIds(ids);
```
Comment (tatianahutii) [inline backend/LeetWars.Core/LeetWars.Core.WebAPI/Controllers/NotificationsController.cs]:
```suggestion
            await _notificationService.UpdateStatusToReadByUserIdsAsync(ids);
```

**Review by tatianahutii — CHANGES_REQUESTED** [review]
See previous comments

File: backend/LeetWars.Core/LeetWars.Core.Common/DTO/Friendship/UpdateFriendshipDto.cs (hunk 9 above)
Comment (AyTee001) [reply backend/LeetWars.Core/LeetWars.Core.Common/DTO/Friendship/UpdateFriendshipDto.cs]: It is needed to work with the analogous enum on the frontend

Comment (tatianahutii) [reply backend/LeetWars.Core/LeetWars.Core.Common/DTO/Friendship/UpdateFriendshipDto.cs]: but we configured the same in out Program file

Comment (AyTee001) [reply backend/LeetWars.Core/LeetWars.Core.Common/DTO/Friendship/UpdateFriendshipDto.cs]: Unfortunately it does not work without explicitly mentioning the converter, i guess something overrides it

Comment (tatianahutii) [reply backend/LeetWars.Core/LeetWars.Core.Common/DTO/Friendship/UpdateFriendshipDto.cs]: strange as it is works for all other places

Comment (AyTee001) [reply backend/LeetWars.Core/LeetWars.Core.Common/DTO/Friendship/UpdateFriendshipDto.cs]: Found the reason, that was SignalR, now it has needed config

**Review by tatianahutii — APPROVED** [review]

**Pull request merged**